### PR TITLE
Increase warning levels

### LINKED ISSUE #428: Increase warning levels
It would be nice to bump warnings up one level to catch a few extra things. Might see a few complaints from 3rd party sources.

msvc: bump to /W4 
gcc: add -Wshadow and -Wconversion (-Wpedantic?)

### PR DESCRIPTION
**Pull request by McCallisterRomer**
resolves #428

- Adding pedantic, shadow, and conversion warnings to GCC.
- Bumping VS warning level up to W4.
- Fixing new found warnings - mostly shadowing + conversion

### CODE REVIEW

File: source/engine/graphics/api/vulkan/shaders/EnvironmentDataShaderResource.cpp [hunk 1]
```diff
@@ -20,7 +20,7 @@ namespace nc::graphics
     {
         auto initialEnvironmentData = EnvironmentData{};
         initialEnvironmentData.cameraWorldPosition = Vector3{-0.0f, 4.0f, -6.4f};
-        initialEnvironmentData.skyboxTextureIndex = -1;
+        initialEnvironmentData.skyboxTextureIndex = 0; // ?
```
Comment (McCallisterRomer) [inline source/engine/graphics/api/vulkan/shaders/EnvironmentDataShaderResource.cpp]: Can you double check this is cool?

File: source/engine/graphics/system/EnvironmentSystem.cpp [hunk 2]
```diff
@@ -41,6 +41,6 @@ auto EnvironmentSystem::Execute(const CameraState& cameraState) -> EnvironmentSt
 void EnvironmentSystem::Clear()
 {
     m_useSkybox = false;
-    m_environmentData.skyboxTextureIndex = -1;
+    m_environmentData.skyboxTextureIndex = 0; // is this ok? why was it -1?
```
Comment (McCallisterRomer) [inline source/engine/graphics/system/EnvironmentSystem.cpp]: Same with this one.

File: source/engine/graphics/api/vulkan/shaders/EnvironmentDataShaderResource.cpp (hunk 1 above)
Comment (jaremieromer) [inline source/engine/graphics/api/vulkan/shaders/EnvironmentDataShaderResource.cpp]: Shoot - we can't actually do this here.

In Pbr.frag, we decision on whether or not to run a block of calculations on whether or not this is >-1.

```    if (environmentData.skyboxCubemapIndex > -1)```
```    {```
 ```       // Environment reflection```
 ```       vec3 I = normalize(inFragPosition - environmentData.cameraWorldPosition);```
 ```       vec3 surfaceNormal = normalize((inNormal + N)/2);```
 ```       vec3 reflected = reflect(I, surfaceNormal);```
 ```       vec3 environmentReflectionColor = SkyboxColor(environmentData.skyboxCubemapIndex, reflected);```

``` result += (F0 * (1-roughnessColor.r) * environmentReflectionColor) / 2;```
```    }```

I suggest changing to ```UINT32_MAX```

File: source/engine/graphics/system/EnvironmentSystem.cpp (hunk 2 above)
Comment (jaremieromer) [inline source/engine/graphics/system/EnvironmentSystem.cpp]: See above comment

File: source/engine/graphics/api/vulkan/shaders/EnvironmentDataShaderResource.cpp (hunk 1 above)
Comment (jaremieromer) [reply source/engine/graphics/api/vulkan/shaders/EnvironmentDataShaderResource.cpp]: See below - an existing bug that needs to be fixed but can't be 0

**Comment by jaremieromer** [comment]
Approved after addressing the existing bug in Environment

**Review by jaremieromer — APPROVED** [review]

File: source/engine/physics/collision/CollisionState.h [hunk 3]
```diff
@@ -8,10 +8,10 @@ namespace nc::physics
 /* Collision data produced by Gjk and consumed by Epa. */
 struct CollisionState
 {
+    DirectX::XMVECTOR rotationA;
```
Comment (McCallisterRomer) [inline source/engine/physics/collision/CollisionState.h:11]: Moved these for alignment.

File: source/engine/graphics/api/vulkan/shaders/EnvironmentDataShaderResource.cpp (hunk 1 above)
Comment (McCallisterRomer) [reply source/engine/graphics/api/vulkan/shaders/EnvironmentDataShaderResource.cpp]: Ok, added an issue and changed to unsigned max

**Pull request merged**